client/homebrew/utils/versionHistory.js

```diff
@@ -0,0 +1,120 @@
+};
+
+function getVersionBySlot(brew, slot){
+    // Read stored brew data
+    // - If it exists, parse data to object
+    // - If it doesn't exist, pass default object
+    const key = getKeyBySlot(brew, slot);
+    const storedVersion = localStorage.getItem(key);
+    const output = storedVersion ? JSON.parse(storedVersion) : { ...DEFAULT_STORED_BREW, ...brew };
+    return output;
+};
+
+function updateStoredBrew(brew, slot=0){
+    const archiveBrew = {};
+
+    // Data from brew to be stored
+    archiveBrew.title = brew.title;
+    archiveBrew.text = brew.text;
+    archiveBrew.style = brew.style;
+    archiveBrew.version = brew.version;
+    archiveBrew.shareId = brew.shareId;
+
+    // Calculated values
+    archiveBrew.savedAt = new Date();
+    archiveBrew.expireAt = new Date();
+    archiveBrew.expireAt.setMinutes(archiveBrew.expireAt.getMinutes() + HISTORY_SAVE_DELAYS[slot]);
+    
+    // Store data
+    const key = getKeyBySlot(brew, slot);
+    localStorage.setItem(key, JSON.stringify(archiveBrew));
+    return;
+};
+
+export function updateHistory(brew) {
+    const numbers = [1,2,3,4,5];
+    const history = {};
+
+    // Load data from local storage to History object
+    numbers.forEach((i)=>{
+        history[i] = getVersionBySlot(brew, i);
+    });
+
+    // Walk each version position
+    numbers.toReversed().every((slot)=>{
+        const storedVersion = history[slot];
+
+        // If slot has expired, update all lower slots and break
+        if(new Date() >= new Date(storedVersion.expireAt)){
+            const slots = Array.from(Array(slot - 1).keys());
+            slots.toReversed().every((slot)=>{
+                const actualSlot = slot + 1;
+                // Move data from actualSlot to actualSlot + 1
+                updateStoredBrew({ ...history[actualSlot], shareId: brew.shareId }, actualSlot + 1);
+                // If first slot, fill with current data
+                if(actualSlot == 1) {
+                    updateStoredBrew(brew, 1);
+                    // Break after updating first slot
+                    return false;
+                }
+                // Continue loop to move data in remaining slots
+                return true;
+            });
+            // Break out of data checks because we found an expired value
+            return false;
+        }
+        // Continue data checks because this one wasn't expired
+        return true;
+    });
+};
+
+export function versionHistoryGarbageCollection(){
+    Object.keys(localStorage)
+        .filter((key)=>{
+            return key.startsWith(HISTORY_PREFIX);
+        })
+        .forEach((key)=>{
+            const collectAt = new Date(JSON.parse(localStorage.getItem(key)).expireAt);
+            collectAt.setMinutes(collectAt.getMinutes() + GARBAGE_COLLECT_DELAY);
+            if(new Date() > collectAt){
+                localStorage.removeItem(key);
+            }
+        });
+};
```

Comment:
I'm not sure I understand the purpose of the garbage collection. Are we just arbitrarily wiping peoples' backups if it's been too long?

Reply:
At this stage, the process only update the entries for the document currently being edited, so once a user completes their brew and never edits it again, the local storage entries persist forever. I'm not sure what the actual limit on local storage is - I seem to recall reading 5MB somewhere - it seemed prudent to have some function that checks ALL entries in local storage and cleans out the ones that have not been altered in a significant period.

Currently I have the non-testing GC expiration value set at 28 days.

Reply:
Digging into the documentation at [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria#web_storage), they list the local storage limit as 5MB.

However, there are other options if this is insufficient.

Reply:
Ok, this is fine then. I just wasn't understanding the motive at first.

Reply:
IndexedDB looks like a really good option - although far more complex than LocalStorage.

Reply:
Yes it does. I say we just do localStorage for now as a minimum viable product and we can switch to indexedDB as its own project.

Reply:
I'll mark this as resolved; currently the garbage collection is only running during the save process of Edit Page, but the intent is that it will be called often.